x/datadeal/abci.go

```diff
@@ -1,13 +1,66 @@
 package datadeal
 
 import (
+	"strconv"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/medibloc/panacea-core/v2/x/datadeal/keeper"
 	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
 	oracletypes "github.com/medibloc/panacea-core/v2/x/oracle/types"
 )
 
 func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+	keeper.IterateClosedDataVerificationQueue(ctx, ctx.BlockHeader().Time, func(dataSale *types.DataSale) bool {
+
+		keeper.RemoveDataVerificationQueue(ctx, dataSale.DealId, dataSale.VerifiableCid, dataSale.VotingPeriod.VotingEndTime)
+		iterator := keeper.GetDataVerificationVoteIterator(ctx, dataSale.DealId, dataSale.VerifiableCid)
+
+		defer iterator.Close()
+
+		oracleKeeper := keeper.GetOracleKeeper()
+
+		tallyResult, err := oracleKeeper.Tally(
+			ctx,
+			iterator,
+			func() oracletypes.Vote {
+				return &types.DataVerificationVote{}
+			},
+			func(vote oracletypes.Vote) error {
+				return keeper.RemoveDataVerificationVote(ctx, vote.(*types.DataVerificationVote))
+			},
+		)
+
+		if err != nil {
+			panic(err)
+		}
+
+		if tallyResult.IsPassed() {
+			dataSale.Status = types.DATA_SALE_STATUS_DELIVERY_VOTING_PERIOD
+			dataSale.VerifiableCid = string(tallyResult.ConsensusValue)
+
+			keeper.AddDataDeliveryQueue(ctx, dataSale.VerifiableCid, dataSale.DealId, oracleKeeper.GetVotingPeriod(ctx).VotingEndTime)
+		} else {
+			dataSale.Status = types.DATA_SALE_STATUS_FAILED
```

Comment:
[As I commented in the previous PR](https://github.com/medibloc/panacea-core/pull/447#discussion_r993030944), I think we need to set this status `DATA_VERIFICATION_FAILED`. In case of failure of delivery vote(whatever the reason is), buyer can re-request delivery to the oracles.

Reply:
Yes, this can be more clearer. So I'll change to `DATA_SALE_STATUS_VERIFICATION_FAILED`.

Reply:
So our `FAILED` statuses have three types. The `VERIFICATION_FAILED`, `DELIVERY_FAILED`, and `FAILED`.
Or is it have two types except `FAILED`?

Reply:
I think it's enough to have two status, `VERIFICATION_FAILED` and `DELIVERY_FAILED`. Do we need a status `FAILED` as well?